README
NAME
    Net::SNMP::Mixin::Dot1qFdb - mixin class for 802.1-Q switch forwarding
    databases

VERSION
    Version 0.04

SYNOPSIS
      use Net::SNMP;
      use Net::SNMP::Mixin qw/mixer init_mixins/;

      my $session = Net::SNMP->session( -hostname => 'foo.bar.com' );
      $session->mixer('Net::SNMP::Mixin::Dot1qFdb');
      $session->init_mixins();
      snmp_dispatcher()   if $session->nonblocking;
      die $session->error if $session->error;

      foreach my $fdb_entry ( $session->get_dot1q_fdb_entries() ) {
        my $mac     = $fdb_entry->{MacAddress};
        my $fdb_id  = $fdb_entry->{fdbId};
        my $vlan_id = $fdb_entry->{vlanId};
        my $port    = $fdb_entry->{dot1dBasePort};
        my $status  = $fdb_entry->{fdbStatus};

        print "$mac, $fdb_id, $vlan_id, $port, $status\n";
      }

DESCRIPTION
    A Net::SNMP mixin class for forwarding database info of 802.1-Q
    compatible switches. The switches must support parts of the standard
    Q-BRIDGE-MIB.

    Sorry to disappoint you, Cisco isn't standard conform, but you knew this
    already, for sure!

MIXIN METHODS
  @fdb = OBJ->get_dot1q_fdb_entries()
    Returns a list of fdb entries. Every list element is a reference to a
    hash with the following fields and values:

        {
          MacAddress      => 'XX:XX:XX:XX:XX:XX',
          dot1dBasePort   => Integer,
          fdbId           => Integer,
          vlanId          => Integer,
          fdbStatus       => Integer,
          fdbStatusString => String,
        }

    MacAddress
        MacAddress received, in normalized IEEE form XX:XX:XX:XX:XX:XX.

    dot1dBasePort
        The receiving bride-port for the MAC address.

    fdbId
        MacAddress is member of the FDB with this fdbId. dot1q bridges
        support many forwarding databases.

    vlanId
        Every fdbId is related to a distinct vlanId.

    fdbStatus
        The status of this entry. The meanings of the values are:

            1 = other
            2 = invalid
            3 = learned
            4 = self
            5 = mgmt

        For more information please see the corresponding Q-BRIDGE-MIB.

    fdbStatusString
        The status of this entry in string form, see above.

INITIALIZATION
  OBJ->_init($reload)
    Fetch the fdb related snmp values from the host. Don't call this method
    direct!

PRIVATE METHODS
    Only for developers or maintainers.

  _fetch_dot1q_fdbid()
    Fetch some columns from the VlanCurrentTable once during object
    initialization. MAC addresses in the forwarding database are related to
    fbd ids and the fbd ids are related to vlan ids by this column.

  _fetch_dot1q_tp_fdb_table()
    Fetch the forwarding databases from the dot1qTpFdbTable once during
    object initialization.

SEE ALSO
    Net::SNMP::Mixin::Dot1dBase for a mapping between ifIndexes and
    dot1dBasePorts.

REQUIREMENTS
    Net::SNMP, Net::SNMP::Mixin

BUGS, PATCHES & FIXES
    There are no known bugs at the time of this release. However, if you
    spot a bug or are experiencing difficulties that are not explained
    within the POD documentation, please submit a bug to the RT system (see
    link below). However, it would help greatly if you are able to pinpoint
    problems or even supply a patch.

    Fixes are dependant upon their severity and my availablity. Should a fix
    not be forthcoming, please feel free to (politely) remind me by sending
    an email to gaissmai@cpan.org .

      RT: http://rt.cpan.org/Public/Dist/Display.html?Name=Net-SNMP-Mixin-Dot1qFdb

AUTHOR
    Karl Gaissmaier <karl.gaissmaier at uni-ulm.de>

COPYRIGHT & LICENSE
    Copyright 2008 Karl Gaissmaier, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Net-SNMP-Mixin-Dot1qFdb

0.04	Fri Sep 26 13:16:57 MEST 2008
	inhibit undefined warnings for wrong table entries

0.03	Fr 25. Apr 19:47:20 CEST 2008
	small internal changes

0.02	Mi 23. Apr 09:24:58 CEST 2008
	First production release
	grml, again Build.PL tweaks

0.01_02 Mi 23. Apr 00:37:02 CEST 2008
        Second developer release.

0.01_01 ??
        First developer release.



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
  module_name       => 'Net::SNMP::Mixin::Dot1qFdb',
  create_readme     => 1,
  license           => 'perl',
  dist_author       => 'Karl Gaissmaier <karl.gaissmaier (at) uni-ulm.de>',
  dist_version_from => 'lib/Net/SNMP/Mixin/Dot1qFdb.pm',
  requires          => {
    'perl'             => '5.6.0',
    'Net::SNMP'        => '0',
    'Net::SNMP::Mixin' => '0',
    'Sub::Exporter'    => '0.978',
  },
  build_requires => {
    'Module::Build' => '0',
    'Test::More'    => '0',
  },
  create_makefile_pl => 'passthrough',
  add_to_cleanup => [qw/Net-SNMP-Mixin-* Makefile* META.yml README/],
);

# -----------------------------------------------------------------------
# shall we perform live tests?

print "==> For running live test you need a network connection <==\n";
if (
  $builder->y_n(
    'Do you want to run live tests against an active SNMP agent?', 'n'
  )
  )
{
  my $snmp_agent =
    $builder->prompt('Enter a hostname/ip-address of snmp agent []:');
  $builder->notes( snmp_agent => $snmp_agent || '' );

  my $community = $builder->prompt('Enter read-only community [public]:');
  $builder->notes( snmp_community => $community || 'public' );

  my $version = $builder->prompt('Enter snmp version [2c]:');
  $builder->notes( snmp_version => $version || '2c' );
}
else {
  print "Skipping live tests.\n";
  $builder->notes( snmp_agent     => '' );
  $builder->notes( snmp_community => '' );
  $builder->notes( snmp_version   => '' );
}

# -----------------------------------------------------------------------

$builder->create_build_script();


examples/qfdb.pl
#!/usr/bin/env perl

use strict;
use warnings;

=head1 NAME

qfdb.pl

=head1 ABSTRACT

A script to get the forwarding database table (FDB) from switches supporting the Q-BRIDGE-MIB:

=head1 SYNOPSIS

 qfdb.pl OPTIONS agent agent ...

 qfdb.pl OPTIONS -i <agents.txt

=head2 OPTIONS

  -c snmp_community
  -v snmp_version
  -t snmp_timeout
  -r snmp_retries

  -d			Net::SNMP debug on
  -i			read agents from stdin, one agent per line
  -B			nonblocking
  -R			print raw FDB table
  -S			print statistics for agents

=head1 DESCRIPTION

Normal output prints the MAC address related to the switch/vlan/port combination with the minimum number of MAC addresses learned. This is a simple but powerful algorithm to find the enduser port or the network insertion port for a given MAC address.

It's also possible to print the whole FDB unprocessed (-R) or summary statistics for any switch/vlan/port (-S).

=head1 REQUIREMENTS

The switches must support the standard Q-BRIDGE-MIB. The script was developed with HP Procurve switches, but any Switch with standard conformity is a good candidate.

Sorry to disappoint you, Cisco isn't standard conform :-(

But you knew this already, for sure!

=cut

use blib;
use Net::SNMP qw(:debug :snmp);
use Net::SNMP::Mixin qw/mixer init_mixins/;
use Getopt::Std;

my %opts;
getopts( 'iRSBdt:r:c:v:', \%opts ) or usage();

my $debug       = $opts{d} || undef;
my $community   = $opts{c} || 'public';
my $version     = $opts{v} || '2';
my $nonblocking = $opts{B} || 0;
my $timeout     = $opts{t} || 10;
my $retries     = $opts{t} || 2;
my $from_stdin  = $opts{i} || undef;
my $print_raw   = $opts{R} || undef;
my $print_stats = $opts{S} || undef;

usage('-R and -S incompatible options') if $opts{R} && $opts{S};

my @agents = @ARGV;
push @agents, <STDIN> if $from_stdin;
chomp @agents;
usage('missing agents') unless @agents;

my @sessions;
foreach my $agent ( sort @agents ) {
  my ( $session, $error ) = Net::SNMP->session(
    -community   => $community,
    -hostname    => $agent,
    -version     => $version,
    -nonblocking => $nonblocking,
    -timeout     => $timeout,
    -retries     => $retries,
    -debug       => $debug ? DEBUG_ALL : 0,
  );

  if ($error) {
    warn $error;
    next;
  }

  $session->mixer(qw/Net::SNMP::Mixin::Dot1qFdb/);
  $session->init_mixins;
  push @sessions, $session;

}
snmp_dispatcher() if $Net::SNMP::NONBLOCKING;

# remove sessions with error from the sessions list
@sessions = grep { warn $_->error if $_->error; not $_->error } @sessions;

# build FDB datastructure over all agents and all macs
# in the first run. We need a second run to decide the
# enduser/insertion port for the MACs
my $fdb    = {};
my $sums   = {};
my $result = {};

build_fdb();

if ($print_raw) {
  print_raw();
  exit 0;
}

if ($print_stats) {
  print_stats();
  exit 0;
}

find_switchport();
print_match();

exit 0;

###################### end of main ######################

sub build_fdb {
  foreach my $session (@sessions) {
    my $agent = $session->hostname;

    foreach my $fdb_entry ( $session->get_dot1q_fdb_entries() ) {
      my $mac        = $fdb_entry->{MacAddress};
      my $vlan_id    = $fdb_entry->{vlanId};
      my $status_str = $fdb_entry->{fdbStatusString};
      my $port       = $fdb_entry->{dot1dBasePort};

      # store the whole knowledge for any fdb_entry in one big hash

      $fdb->{$vlan_id}{$mac}{$agent}{status_str} = $status_str;
      $fdb->{$vlan_id}{$mac}{$agent}{port}       = $port;

      # count the macs per agent/vlan/port in order to decide
      # if this is an enduser port or a trunk

      $sums->{macs}{$agent}++;
      $sums->{vlan}{$agent}{$vlan_id}++;
      $sums->{port}{$agent}{$vlan_id}{$port}++;

    }
  }
}

# find the agent and port with the less connected MACs
# stuff it into result hash
sub find_switchport {

  foreach my $vlan_id ( keys %$fdb ) {
    my @macs = keys %{ $fdb->{$vlan_id} };
    foreach my $mac (@macs) {

      my @agents = keys %{ $fdb->{$vlan_id}{$mac} };
      foreach my $agent (@agents) {

        # get port and status for current mac in this vlan on this agent
        my $port = $fdb->{$vlan_id}{$mac}{$agent}{port};
        next unless defined $port;

        my $status_str = $fdb->{$vlan_id}{$mac}{$agent}{status_str};
        next unless defined $status_str;

        # check if we find a port with less connected systems,
        # this is a candidate for this MAC as enduser port

        my $this_port_sum = $sums->{port}{$agent}{$vlan_id}{$port};
        next unless $this_port_sum > 0;

        my $min_port_sum = $result->{$vlan_id}{$mac}{min_port_sum};

        # when do we replace the current best entry?
        if (
          not defined $min_port_sum or    # first match
          $port == 0                or    # the switch own VLAN MAC address
          $this_port_sum < $min_port_sum  # better match
          )
        {

          # this is a candidate for this MAC as enduser port
          $result->{$vlan_id}{$mac}{min_port_sum} = $this_port_sum;
          $result->{$vlan_id}{$mac}{agent}        = $agent;
          $result->{$vlan_id}{$mac}{port}         = $port;
          $result->{$vlan_id}{$mac}{status_str}   = $status_str;

        }
      }
    }
  }
}

# print the best match switchport for a MAC
sub print_match {

  # sort by vlan_id
  foreach my $vlan_id ( sort { $a <=> $b } keys %$result ) {

    # sort by port
    foreach my $mac (
      sort {
        $result->{$vlan_id}{$a}{port} <=> $result->{$vlan_id}{$b}{port}
      }
      keys %{ $result->{$vlan_id} }
      )
    {
      my $port         = $result->{$vlan_id}{$mac}{port};
      my $status_str   = $result->{$vlan_id}{$mac}{status_str};
      my $agent        = $result->{$vlan_id}{$mac}{agent};
      my $min_port_sum = $result->{$vlan_id}{$mac}{min_port_sum};

      printf
        "%-5d bridge_port(%3d) vlan(%4d) mac(%s) status(%7s) agent(%s)\n",
        $min_port_sum, $port, $vlan_id, $mac, $status_str, $agent;
    }
    print "\n";
  }
}

sub print_raw {

  # resort for print order vlan_id, port, agent
  my $resort = {};

  foreach my $vlan_id ( keys %$fdb ) {

    my @macs = keys %{ $fdb->{$vlan_id} };
    foreach my $mac (@macs) {

      my @agents = keys %{ $fdb->{$vlan_id}{$mac} };
      foreach my $agent ( sort @agents ) {

        # get port and status for current mac in this vlan on this agent
        my $port = $fdb->{$vlan_id}{$mac}{$agent}{port};
        next unless defined $port;

        my $status_str = $fdb->{$vlan_id}{$mac}{$agent}{status_str};
        next unless defined $status_str;

        my $this_port_sum = $sums->{port}{$agent}{$vlan_id}{$port};
        next unless $this_port_sum > 0;

        # store FDB in other sort order for raw printing
        $resort->{$agent}{$vlan_id}{$port}{$mac}{status_str} = $status_str;
      }
    }
  }

  # now print in resort order
  foreach my $agent ( sort keys %$resort ) {

    my @vlans = sort { $a <=> $b } keys %{ $resort->{$agent} };
    foreach my $vlan_id (@vlans) {

      my @ports = sort { $a <=> $b } keys %{ $resort->{$agent}{$vlan_id} };
      foreach my $port (@ports) {

        my @macs = sort keys %{ $resort->{$agent}{$vlan_id}{$port} };
        foreach my $mac (@macs) {

          my $status_str =
            $resort->{$agent}{$vlan_id}{$port}{$mac}{status_str};
          my $port_sum = $sums->{port}{$agent}{$vlan_id}{$port};

          printf
"%-5d bridge_port(%3d) vlan(%4d) mac(%s) status(%7s) agent(%s)\n",
            $port_sum, $port, $vlan_id, $mac, $status_str, $agent;
        }
      }
      print "\n";
    }
    print "\n";
  }
}

sub print_stats {

  foreach my $agent ( sort keys %{ $sums->{vlan} } ) {

    my @vlans = sort { $a <=> $b } keys %{ $sums->{vlan}{$agent} };
    foreach my $vlan_id (@vlans) {

      my $vlan_sum = $sums->{vlan}{$agent}{$vlan_id};

      printf "agent(%s) vlan(%4d) addresses(%4d)\n", $agent, $vlan_id,
        $vlan_sum;

    }
    print "\n";
  }
}

sub usage {
  my @msg = @_;
  die <<EOT;
>>>>>> @msg
    Usage: $0 [options] hostname
   
    	-c community
  	-v version
  	-t timeout
  	-r retries
  	-d		Net::SNMP debug on
	-i		read agents from stdin
  	-B		nonblocking
  	-R		print raw FDB table
  	-S		print statistics for agents
EOT
}

=head1 AUTHOR

Karl Gaissmaier, karl.gaissmaier (at) uni-ulm.de

=head1 COPYRIGHT

Copyright (C) 2008 by Karl Gaissmaier

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

# vim: sw=2


lib/Net/SNMP/Mixin/Dot1qFdb.pm
package Net::SNMP::Mixin::Dot1qFdb;

use strict;
use warnings;

#
# store this package name in a handy variable,
# used for unambiguous prefix of mixin attributes
# storage in object hash
#
my $prefix = __PACKAGE__;

#
# this module import config
#
use Carp ();
use Net::SNMP::Mixin::Util qw/idx2val normalize_mac/;

#
# this module export config
#
my @mixin_methods;

BEGIN {
  @mixin_methods = ( qw/ get_dot1q_fdb_entries /);
}

use Sub::Exporter -setup => {
  exports => [@mixin_methods],
  groups  => { default => [@mixin_methods], },
};

#
# SNMP oid constants used in this module
#
use constant {
  DOT1Q_TP_FDB_TABLE   => '1.3.6.1.2.1.17.7.1.2.2',
  DOT1Q_TP_FDB_ADDRESS => '1.3.6.1.2.1.17.7.1.2.2.1.1',
  DOT1Q_TP_FDB_PORT    => '1.3.6.1.2.1.17.7.1.2.2.1.2',
  DOT1Q_TP_FDB_STATUS  => '1.3.6.1.2.1.17.7.1.2.2.1.3',

  DOT1Q_VLAN_CURRENT_FDB_ID => '1.3.6.1.2.1.17.7.1.4.2.1.3',
};

=head1 NAME

Net::SNMP::Mixin::Dot1qFdb - mixin class for 802.1-Q switch forwarding databases

=head1 VERSION

Version 0.04

=cut

our $VERSION = '0.04';

=head1 SYNOPSIS

  use Net::SNMP;
  use Net::SNMP::Mixin qw/mixer init_mixins/;

  my $session = Net::SNMP->session( -hostname => 'foo.bar.com' );
  $session->mixer('Net::SNMP::Mixin::Dot1qFdb');
  $session->init_mixins();
  snmp_dispatcher()   if $session->nonblocking;
  die $session->error if $session->error;

  foreach my $fdb_entry ( $session->get_dot1q_fdb_entries() ) {
    my $mac     = $fdb_entry->{MacAddress};
    my $fdb_id  = $fdb_entry->{fdbId};
    my $vlan_id = $fdb_entry->{vlanId};
    my $port    = $fdb_entry->{dot1dBasePort};
    my $status  = $fdb_entry->{fdbStatus};

    print "$mac, $fdb_id, $vlan_id, $port, $status\n";
  }

=head1 DESCRIPTION

A Net::SNMP mixin class for forwarding database info of 802.1-Q compatible switches. The switches must support parts of the standard Q-BRIDGE-MIB.

Sorry to disappoint you, Cisco isn't standard conform, but you knew this already, for sure!

=head1 MIXIN METHODS

=head2 B<< @fdb = OBJ->get_dot1q_fdb_entries() >>

Returns a list of fdb entries. Every list element is a reference to a hash with the following fields and values:

    {
      MacAddress      => 'XX:XX:XX:XX:XX:XX',
      dot1dBasePort   => Integer,
      fdbId           => Integer,
      vlanId          => Integer,
      fdbStatus       => Integer,
      fdbStatusString => String,
    }

=over

=item MacAddress

MacAddress received, in normalized IEEE form XX:XX:XX:XX:XX:XX.

=item dot1dBasePort

The receiving bride-port for the MAC address.

=item fdbId

MacAddress is member of the FDB with this fdbId. dot1q bridges support many forwarding databases.

=item vlanId

Every fdbId is related to a distinct vlanId.

=item fdbStatus

The status of this entry. The meanings of the values are:

    1 = other
    2 = invalid
    3 = learned
    4 = self
    5 = mgmt

For more information please see the corresponding Q-BRIDGE-MIB.

=item fdbStatusString

The status of this entry in string form, see above.

=back

=cut

sub get_dot1q_fdb_entries {
  my $session = shift;
  Carp::croak "'$prefix' not initialized,"
    unless $session->{$prefix}{__initialized};

  #
  # the port's current state translation table
  #
  my %fdp_entry_status_enum = (
    1 => 'other',
    2 => 'invalid',
    3 => 'learned',
    4 => 'self',
    5 => 'mgmt',
  );

  # stash for return values
  my @fdb_entries = ();

  my ( @digits, $fdb_id, $vlan_id, $mac, $mac_string, $port, $status,
    $status_string );

  # index is fdbId.MacAddress
  foreach my $idx ( keys %{ $session->{$prefix}{dot1qTpFdbPort} } ) {
    $port   = $session->{$prefix}{dot1qTpFdbPort}{$idx};
    $status = $session->{$prefix}{dot1qTpFdbStatus}{$idx};

    # the snmp get_table() isn't a snapshot, it can be, that
    # the MAC has already timeout in the FDB when the
    # status is fetched
    next unless defined $port && defined $status;

    $status_string = $fdp_entry_status_enum{$status};

    # split the idx to fdb_id and mac address
    # index is fdbId.MacAddress, value is the bridge port
    @digits = split /\./, $idx;

    $fdb_id = $digits[0];
    $vlan_id = $session->{$prefix}{fdb_id2vlan_id}{$fdb_id};

    $mac        = pack( 'C6', @digits[ 1 .. 6 ] );
    $mac_string = normalize_mac($mac);

    push @fdb_entries,
      {
      dot1dBasePort   => $port,
      MacAddress      => $mac_string,
      fdbId           => $fdb_id,
      vlanId          => $vlan_id,
      fdbStatus       => $status,
      fdbStatusString => $status_string,
      };
  }

  return @fdb_entries;
}

=head1 INITIALIZATION

=cut

=head2 B<< OBJ->_init($reload) >>

Fetch the fdb related snmp values from the host. Don't call this method direct!

=cut

sub _init {
  my ( $session, $reload ) = @_;

  die "$prefix already initalized and reload not forced.\n"
    if $session->{$prefix}{__initialized} && not $reload;

  # initialize the object for forwarding databases infos
  _fetch_dot1q_fdbid($session);
  return if $session->error;

  _fetch_dot1q_tp_fdb_entries($session);
  return if $session->error;

  return 1;
}

=head1 PRIVATE METHODS

Only for developers or maintainers.

=head2 B<< _fetch_dot1q_fdbid() >>

Fetch some columns from the VlanCurrentTable once during object initialization. MAC addresses in the forwarding database are related to fbd ids and the fbd ids are related to vlan ids by this column.

=cut

sub _fetch_dot1q_fdbid() {
  my $session = shift;
  my $result;

  # fetch the dot1qVlanFdbId from dot1qVlanCurrentTable
  $result = $session->get_table(
    -baseoid => DOT1Q_VLAN_CURRENT_FDB_ID,

    # define callback if in nonblocking mode
    $session->nonblocking ? ( -callback => \&_dot1q_fdbid_cb ) : (),
  );

  return unless defined $result;
  return 1 if $session->nonblocking;

  # call the callback function in blocking mode by hand
  _dot1q_fdbid_cb($session);

}

sub _dot1q_fdbid_cb {
  my $session = shift;
  my $vbl     = $session->var_bind_list;

  return unless defined $vbl;

  # mangle result table to get plain vlan_id => fdb_id
  #
  #         1.3.6.1.2.1.17.7.1.4.2.1.3.0.n => m
  #                                  | | |    |
  # DOT1Q_VLAN_CURRENT_FDB_ID -------/ | |    |
  #         dot1qVlanTimeMark ---------/ |    |
  #            dot1qVlanIndex -----------/    |
  #            dot1qVlanFdbId ----------------/

  # vlan_id => fdb_id
  $session->{$prefix}{vlan_id2fdb_id} =
    idx2val( $vbl, DOT1Q_VLAN_CURRENT_FDB_ID, 1 );

  # build reverse map fdb_id => vlan_id
  while ( my ( $vlan_id, $fdb_id ) =
    each %{ $session->{$prefix}{vlan_id2fdb_id} } )
  {

    $session->{$prefix}{fdb_id2vlan_id}{$fdb_id} = $vlan_id;
  }

  $session->{$prefix}{__initialized}++;
}

=head2 B<< _fetch_dot1q_tp_fdb_table() >>

Fetch the forwarding databases from the dot1qTpFdbTable once during object initialization.

=cut

sub _fetch_dot1q_tp_fdb_entries() {
  my $session = shift;
  my $result;

  # fetch the forwarding databases from dot1qTpFdbTable
  $result = $session->get_entries(
    -columns => [ DOT1Q_TP_FDB_PORT, DOT1Q_TP_FDB_STATUS ],

    # define callback if in nonblocking mode
    $session->nonblocking
    ? ( -callback => \&_dot1q_tp_fdb_entries_cb )
    : (),
  );

  return unless defined $result;
  return 1 if $session->nonblocking;

  # call the callback function in blocking mode by hand
  _dot1q_tp_fdb_entries_cb($session);

}

sub _dot1q_tp_fdb_entries_cb {
  my $session = shift;
  my $vbl     = $session->var_bind_list;

  return unless defined $vbl;

  # mangle result table to get plain idx->value
  # index is fdbId.MacAddress, value is the bridge port
  $session->{$prefix}{dot1qTpFdbPort} = idx2val( $vbl, DOT1Q_TP_FDB_PORT );

  # mangle result table to get plain idx->value
  # index is fdbId.MacAddress, value is the entry status
  $session->{$prefix}{dot1qTpFdbStatus} = idx2val( $vbl, DOT1Q_TP_FDB_STATUS );

  $session->{$prefix}{__initialized}++;
}

=head1 SEE ALSO

L<< Net::SNMP::Mixin::Dot1dBase >> for a mapping between ifIndexes and dot1dBasePorts.

=head1 REQUIREMENTS

L<< Net::SNMP >>, L<< Net::SNMP::Mixin >>

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a bug or are experiencing difficulties that are not explained within the POD documentation, please submit a bug to the RT system (see link below). However, it would help greatly if you are able to pinpoint problems or even supply a patch. 

Fixes are dependant upon their severity and my availablity. Should a fix not be forthcoming, please feel free to (politely) remind me by sending an email to gaissmai@cpan.org .

  RT: http://rt.cpan.org/Public/Dist/Display.html?Name=Net-SNMP-Mixin-Dot1qFdb

=head1 AUTHOR

Karl Gaissmaier <karl.gaissmaier at uni-ulm.de>

=head1 COPYRIGHT & LICENSE

Copyright 2008 Karl Gaissmaier, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

unless ( caller() ) {
  print "$prefix compiles and initializes successful.\n";
}

1;

# vim: sw=2
